### ISSUE TITLE
Failed to install Openshift 3.10 because of missing packages

### ISSUE THREAD

**Issue by gbenhaim** [openshift/openshift-ansible#8399]
#### Description

Failed to install Openshift 3.10 because of missing packages.

##### Version

```
ansible 2.5.2
openshift-ansible-3.10.0-0.41.0
OS - Centos 7.5
```

##### Steps To Reproduce

Run `deploy_cluster.yml` playbook

##### Expected Results
```
Openshift origin 3.10 installed successfully
```

##### Observed Results
```
TASK [openshift_node : Install node, clients, and conntrack packages] **********
Wednesday 16 May 2018  15:57:47 +0000 (0:00:02.764)       0:10:00.906 ********* 
FAILED - RETRYING: Install node, clients, and conntrack packages (3 retries left).
FAILED - RETRYING: Install node, clients, and conntrack packages (3 retries left).
FAILED - RETRYING: Install node, clients, and conntrack packages (3 retries left).
FAILED - RETRYING: Install node, clients, and conntrack packages (2 retries left).
FAILED - RETRYING: Install node, clients, and conntrack packages (2 retries left).
FAILED - RETRYING: Install node, clients, and conntrack packages (2 retries left).
FAILED - RETRYING: Install node, clients, and conntrack packages (1 retries left).
FAILED - RETRYING: Install node, clients, and conntrack packages (1 retries left).
FAILED - RETRYING: Install node, clients, and conntrack packages (1 retries left).
failed: [lago-node0] (item={u'name': u'origin-node-3.10*'}) => {
    "attempts": 3, 
    "changed": false, 
    "item": {
        "name": "origin-node-3.10*"
    }, 
    "rc": 126, 
    "results": [
        "No package matching 'origin-node-3.10*' found available, installed or updated"
    ]
}
```
```
No package matching 'origin-clients-3.10*' found available, installed or updated
failed: [lago-node0] (item={u'name': u'origin-clients-3.10*'}) => {
    "attempts": 3, 
    "changed": false, 
    "item": {
        "name": "origin-clients-3.10*"
    }, 
    "rc": 126, 
    "results": [
        "No package matching 'origin-clients-3.10*' found available, installed or updated"
    ]
}
```

##### Additional Information

* Centos 7.5
* https://gist.github.com/gbenhaim/1f816581e2b1d3d0db2dd7c6796d3ddd

- I also didn't see that origin-*-3.10* is available in the testing repo:
https://buildlogs.centos.org/centos/7/paas/x86_64/openshift-origin/

- I don't understand why installing any RPM packages on the masters/nodes when `containerized=true`

**Comment by DanyC97** [comment]
@gbenhaim correct the rpms for 3.10 are not yet built, am planning to work on it this week however it won't be published on official repo in the next 2 week. Once will be available as candidate i'll tag you, sorry for delay

**Comment by gbenhaim** [comment]
@DanyC97 thanks.

**Comment by gbenhaim** [comment]
@nellyc

**Comment by gbenhaim** [comment]
@DanyC97 any updates?

**Comment by DanyC97** [comment]
@gbenhaim sorry no, just got back to work. is on my top priority list, just please bear with me a bit.

**Comment by ticapix** [comment]
@gbenhaim : any update or way we could help ?

**Comment by gbenhaim** [comment]
@ticapix Is there any official repo for 3.10 rc?

**Comment by moula** [comment]
@DanyC97 Repo 3.10 for soon or not yet? Thank's

**Comment by rhnewtron** [comment]
Any news here. Openshift 3.10 has been released by RedHat, however origin is not available

**Comment by DanyC97** [comment]
guys i understand the excitement but it doesn't help if everyone is adding a comment on top of another comment.

the reason i haven't released as part of PaaS SIG any origin rpm for v3.10.rc0 is because we couldn't build it due to missing code which was made in OCP but has been ported to Origin but that is now [sorted](https://github.com/openshift/origin/pull/20503) 

Once that was done i was waiting for the new tag in Origin repo to be created which was done few hours ago with the official v3.10 release and so i'm working on it to get the rpms out for testing.

**Comment by DanyC97** [comment]
in case you haven't seen the [notification](http://lists.openshift.redhat.com/openshift-archives/dev/2018-August/msg00000.html) sent to dev mailing list 

Let me know if there are any issues either here or writing to the mailer

**Comment by ticapix** [comment]
The default `playbooks/prerequisites.yml` and `playbooks/deploy_cluster.yml` are working on the first try on a fresh Centos7 install with the default `inventory/hosts.localhost` config.
The only tweak was

```shell
pip uninstall ansible || true
yum -y install yum # this will install ansible-2.4
pip install --upgrade ansible  # this will install ansible-2.6

```

Otherwise if using ansible-2.6 installed from pip and without having the package installed from the distrib package manager, the playbook, will force the install of 2.4 version and it messes up the script at some point. (error about environment 'shell' var not being defined)

**Comment by DanyC97** [comment]
eh ?

to my knowledge 3.10 is working with 2.4, @sdodson am i missing something here ?
only 3.11 will require 2.6 ansible 

You saying that deploying 3.10 with ansible 2.4 doesn't work ?

**Comment by ticapix** [comment]
I'm saying that if ansible 2.6 is installed via pip before running the
playbook it fails. I'll try with the default 2.4 version from yum.

Le ven. 3 août 2018 à 20:19, Dani Comnea <notifications@github.com> a
écrit :

**Comment by DanyC97** [comment]
right which is what i expected @ticapix .
As mentioned 2.6 is only for 3.11 which is not released, is work in progress in master(dev) branch.

**Comment by imranrazakhan** [comment]
I just installed but still getting below error

```
git clone https://github.com/openshift/openshift-ansible.git
cd openshift-ansible && git fetch && git checkout release-3.10 && cd ..
ansible-playbook -i inventory.ini openshift-ansible/playbooks/prerequisites.yml
ansible-playbook -i inventory.ini openshift-ansible/playbooks/deploy_cluster.yml
```
Following is error

`No package matching 'origin-clients-3.10*' found available, installed or updated`

**Comment by DanyC97** [comment]
@imranrazakhan which yum repo have you used ?

**Comment by isaldarriaga** [comment]
@imranrazakhan production repos won't work yet.

check dev mailing list for testing RPMs: [http://lists.openshift.redhat.com/openshift-archives/dev/2018-August/msg00000.html](url)

Anyone to post how to activate them?

**Comment by ticapix** [comment]
```
cat > /etc/yum.repos.d/CentOS-OpenShift-Origin-CBS.repo <<EOF
[centos-openshift-origin-testing-cbs]
name=CentOS OpenShift Origin Testing CBS
baseurl=https://cbs.centos.org/repos/paas7-openshift-origin310-testing/x86_64/os/
enabled=1
gpgcheck=0
gpgkey=file:///etc/pki/rpm-gpg/openshift-ansible-CentOS-SIG-PaaS
EOF
```

**Comment by DanyC97** [comment]
@ticapix and you saying with the above repo things still don't work ?

**Comment by chrisop** [comment]
@DanyC97 @imranrazakhan 
For my 3.10 installation I used the following baseurl in the .repo file:
https://cbs.centos.org/repos/paas7-openshift-origin310-candidate/x86_64/os/Packages/

The installation of the package origin-node-3.10.0 worked without problems.
I used for the installation the Ansible version 2.6.2. and the openshift-ansible repo with the branch release-3.10. 
Only the openshift-infra deployment failed because of image pull errors of version v3.10, it seems not be build already. Version latest worked for my test without problems.

**Comment by Reamer** [comment]
@DanyC97 I was able to upgrade to version 3.10 with the provided rpms. Is there anything I can do for you to help you to get this RPM into production?
@isaldarriaga At the moment it's **not** possibly to activate this repo with openshift-ansible 
@ticapix Thanks for your repo definition. I copied this definition via ansible ad hoc command to my cluster nodes

**Comment by DanyC97** [comment]
@Reamer thank you kindly for your help, there is nothing major left now that you confirmed. I just sent out another [notification](http://lists.openshift.redhat.com/openshift-archives/dev/2018-August/msg00001.html) with the next steps. 

sorry for delay but not everything is under my control.

**Comment by isaldarriaga** [comment]
@Reamer FYI i set the repo in all host as per @ticapix instructions and ansible was able to pull the images without problem.

My installation is not working though.

i traced down the error to the api control plane pod:

`$ master-logs api $(docker ps | grep api | awk '{print $1}')`
**json: cannot unmarshal array into Go value of type types.ContainerJSON**

the pod's yaml looks fine, but not quite sure if causing the problem or not:

`$ cat /etc/origin/node/pods/apiserver.yaml`

apiVersion: v1
kind: Pod
metadata:
  annotations:
    scheduler.alpha.kubernetes.io/critical-pod: ''
  labels:
    openshift.io/component: api
    openshift.io/control-plane: 'true'
  name: master-api
  namespace: kube-system
spec:
  containers:
  - args:
    - "#!/bin/bash\nset -euo pipefail\nif [[ -f /etc/origin/master/master.env ]];\
      \ then\n  set -o allexport\n  source /etc/origin/master/master.env\nfi\nexec\
      \ openshift start master api --config=/etc/origin/master/master-config.yaml\
      \ --loglevel=${DEBUG_LOGLEVEL:-2}\n"
    command:
    - /bin/bash
    - -c
    image: docker.io/openshift/origin-control-plane:v3.10.0
    livenessProbe:
      httpGet:
        path: healthz
        port: 443
        scheme: HTTPS
      initialDelaySeconds: 45
      timeoutSeconds: 10
    name: api
    readinessProbe:
      httpGet:
        path: healthz/ready
        port: 443
        scheme: HTTPS
      initialDelaySeconds: 10
      timeoutSeconds: 10
    securityContext:
      privileged: true
    volumeMounts:
    - mountPath: /etc/origin/master/
      name: master-config
    - mountPath: /etc/origin/cloudprovider/
      name: master-cloud-provider
    - mountPath: /var/lib/origin/
      name: master-data
  hostNetwork: true
  restartPolicy: Always
  volumes:
  - hostPath:
      path: /etc/origin/master/
    name: master-config
  - hostPath:
      path: /etc/origin/cloudprovider
    name: master-cloud-provider
  - hostPath:
      path: /var/lib/origin
    name: master-data`

docker logs shows nothing.

i couldn't find anything in /var/log for /usr/bin/pod in the api control plane's container either:

`$ docker exec -it $(docker ps | grep api | awk '{print $1}') ps aux`
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
1001         1  0.0  0.0   2764   720 ?        Ssl  Aug07   0:00 /usr/bin/pod
1001        44  2.0  0.0  51708  1652 ?        Rs+  00:57   0:00 ps aux

`$ docker exec -it $(docker ps | grep api | awk '{print $1}') ls /var/log`
btmp  grubby_prune_debug  lastlog  tallylog  wtmp  yum.log

how to increase the verbosity in master-logs / docker container? already is 8 in ansible and control plane looks already set to that.

`$ cat /etc/origin/master/master.env`

# Proxy configuration
# See https://docs.openshift.com/enterprise/latest/install_config/install/advanced_install.html#configuring-global-proxy

DEBUG_LOGLEVEL=8

This is the docker's config just in case:

`$ cat /etc/sysconfig/docker`
# /etc/sysconfig/docker

# Modify these options if you want to change the way the docker daemon runs
OPTIONS=' --selinux-enabled       --signature-verification=False'
if [ -z "${DOCKER_CERT_PATH}" ]; then
    DOCKER_CERT_PATH=/etc/docker
fi

# Do not add registries in this file anymore. Use /etc/containers/registries.conf
# from the atomic-registries package.
#

# Location used for temporary files, such as those created by
# docker load and build operations. Default is /var/lib/docker/tmp
# Can be overriden by setting the following environment variable.
# DOCKER_TMPDIR=/var/tmp

# Controls the /etc/cron.daily/docker-logrotate cron job status.
# To disable, uncomment the line below.
# LOGROTATE=false

# docker-latest daemon can be used by starting the docker-latest unitfile.
# To use docker-latest client, uncomment below lines
#DOCKERBINARY=/usr/bin/docker-latest
#DOCKERDBINARY=/usr/bin/dockerd-latest
#DOCKER_CONTAINERD_BINARY=/usr/bin/docker-containerd-latest
#DOCKER_CONTAINERD_SHIM_BINARY=/usr/bin/docker-containerd-shim-latest

Inventory file attached:
[inventory.txt](https://github.com/openshift/openshift-ansible/files/2268556/inventory.txt)

ansible log (verbosity = 8)
[ansible-2018-08-07T18:32:37.zip](https://github.com/openshift/openshift-ansible/files/2268559/ansible-2018-08-07T18.32.37.zip)

**Comment by Reamer** [comment]
@isaldarriaga It was very helpful to start the openshift api service manual with `openshift start master api --config=/etc/origin/master/master-config.yaml --loglevel=${DEBUG_LOGLEVEL:-2}`
Then you should be able to look via webconsole into the output log of your failed pods in namespace `kube-system`. In my case it was an old vsphere config file `/etc/vsphere/vsphere.conf` - custom stuff...

**Comment by isaldarriaga** [comment]
@Reamer i started the api that way.

`$ openshift start master api --config=/etc/origin/master/master-config.yaml --loglevel=8`
[https://gist.github.com/isaldarriaga/0f9e0a536ae6c414951609d9c2d8c877](https://gist.github.com/isaldarriaga/0f9e0a536ae6c414951609d9c2d8c877)

The $DEBUG_LOGLEVEL is empty by default:
`$ env`
[https://gist.github.com/isaldarriaga/9ff9da4c8e842d2f4aabd6f7dc28093d](https://gist.github.com/isaldarriaga/9ff9da4c8e842d2f4aabd6f7dc28093d)

`$ cat /etc/origin/master/master-config.yaml`
[https://gist.github.com/isaldarriaga/975f86539c12825de643d4944547d625](https://gist.github.com/isaldarriaga/975f86539c12825de643d4944547d625)

`openshift start master api` command's last output is :
`Invalid MasterConfig /etc/origin/master/master-config.yaml`
`  etcdClientInfo.ca: Invalid value: "/etc/origin/master/master.etcd-ca.crt": could not read file: stat /etc/origin/master/master.etcd-ca.crt: no such file or directory`

`$ ls -al /etc/origin/master/`
[https://gist.github.com/isaldarriaga/6bd94aec66473a15e1315a57519b2ded](https://gist.github.com/isaldarriaga/6bd94aec66473a15e1315a57519b2ded)

Since the api didn't come up, i cannot get anything from oc:

`$ oc get pods -n kube-system`
`The connection to the server prod-master-01.example.com:443 was refused - did you specify the right host or port?`

This is what i attempted:

a) `$ cp /etc/origin/master/ca-bundle.crt /etc/origin/master/ca-bundle.crt && openshift start master api --config=/etc/origin/master/master-config.yaml --loglevel=8`
. . .
`F0808 06:07:12.757726   25569 start_api.go:68] dial tcp 192.168.246.111:2379: getsockopt: connection refused`

the master-01 (.111) attempts to reach itself over :2379

b) `$ cp /etc/origin/master/ca.crt /etc/origin/master/client-ca-bundle.crt && openshift start master api --config=/etc/origin/master/master-config.yaml --loglevel=8`
. . .
`F0808 06:13:58.080386   26172 start_api.go:68] dial tcp 192.168.246.112:2379: getsockopt: no route to host`

the master-01 (.111) attempts to reach master-02 (.112) over :2379

`$ iptables --list-rules`
[https://gist.github.com/isaldarriaga/eba36ac855ce37844c7f0b10eac8615f](https://gist.github.com/isaldarriaga/eba36ac855ce37844c7f0b10eac8615f)

c) `$ cp /etc/origin/master/ca.crt /etc/origin/master/master.etcd-ca.crt && openshift start master api --config=/etc/origin/master/master-config.yaml --loglevel=8`
. . .
`F0808 05:54:39.059864   24454 start_api.go:68] dial tcp 192.168.246.112:2379: getsockopt: no route to host`

the master-01 (.111) attempts to reach master-02 (.112) over :2379

d) `$ cp /etc/origin/master/ca.crt /etc/origin/master/frontproxy-ca.crt && openshift start master api --config=/etc/origin/master/master-config.yaml --loglevel=8`
. . .
`F0808 06:26:58.111176   27244 start_api.go:68] dial tcp 192.168.246.113:2379: getsockopt: no route to host`

the master-01 (.111) attempts to reach master-03 (.113) over :2379


so there is a number of problems:
1. DEBUG_LOGLEVEL env is not populated during installation, affecting `openshift start master api` verbosity
2. /etc/origin/master/master.etcd-ca.crt is not generated before starting up api pod. Though other certs are.
3. master-logs api $(docker ps | grep api | awk '{print $1}') command's output does not correspond to 
`openshift start master api` command's output. maybe caused by #1.
4. iptables :2379 rule is missing in all masters

**Comment by Reamer** [comment]
@DanyC97 This has nothing to do with missing packages. 
@isaldarriaga Please open a new issue for your problem.

**Comment by DanyC97** [comment]
okay @Reamer , thank you for confirmation.  But can you confirm if you you managed to get a working/ running 3.10 cluster ?

**Comment by Reamer** [comment]
I can confirm that i have now a running 3.10 cluster with your provided rpms using the [openshift-ansible upgrade playbook](https://github.com/openshift/openshift-ansible/blob/release-3.10/playbooks/byo/openshift-cluster/upgrades/v3_10/upgrade.yml).

**Comment by DanyC97** [comment]
thank you v much for confirmation @Reamer !
@gbenhaim can you please close this issue that that the rpms been provided  ? i'll complete the final step on promoting to mirror.centos in the next few hours and it should get in sync by tmw.

**Comment by isaldarriaga** [comment]
@DanyC97, i detected one missing rpm so far (python-docker).

on TASK [Ensure openshift-ansible installer package deps are installed]:
`"item": "python-docker", `
`    "msg": "Failure talking to yum: failure: repodata/repomd.xml from centos-openshift-origin-testing-cbs: [Errno 256] No more mirrors to try.\nhttps://cbs.centos.org/repos/paas7-openshift-origin310-testing/x86_64/os/Packages/repodata/repomd.xml: [Errno 14] HTTPS Error 404 - Not Found"`

there's no such rpm in any of the 3 referenced openshift repos.

but here:
https://rpmfind.net/linux/rpm2html/search.php?query=python-docker-py

I reinstalled having etcd in separate hosts this time.

inventory:
[https://gist.github.com/isaldarriaga/43ed5c19a32f4b2ed778a7b59952439a](https://gist.github.com/isaldarriaga/43ed5c19a32f4b2ed778a7b59952439a)

install logs (verbose = 8)
[ansible.zip](https://github.com/openshift/openshift-ansible/files/2272705/ansible.zip)

**Comment by DanyC97** [comment]
@isaldarriaga the error you showing me is related to missing the `python-docker` rpm but then you are showing me in `rpmfind.net` the `python-docker-py` package .

Now looking in the code i can see [this](https://github.com/openshift/openshift-ansible/blob/release-3.10/playbooks/init/base_packages.yml#L42) so i suspect it was a typo ?

**Comment by DanyC97** [comment]
and for your info @isaldarriaga i just tagged `python-docker-3.3.0-1.el7` into [testing](https://cbs.centos.org/repos/paas7-openshift-origin310-testing/x86_64/os/Packages/) repo, any chance you can give it another run  ?

**Comment by isaldarriaga** [comment]
you're right. i was trying to point out to the existence of the package, but typo.

My lab's hosts is centos7, my client fedora28 (ansible runs here).

Does the code mean if ansible running in fedora, no matter the host's OS, it should install python-docker instead of python3-docker?

now seeing python2-docker-3.3.0-1.el7.noarch.rpm in centos CBS/buildlogs.

thanks, i'm going to retry.

**Comment by isaldarriaga** [comment]
ok. Some dependencies are missing when `yum -y update`:

Error: Package: python2-docker-3.3.0-1.el7.noarch (centos-openshift-origin-testing-cbs)
           Requires: python-requests >= 2.14.2
           Installed: python-requests-2.6.0-1.el7_1.noarch (@base)
               python-requests = 2.6.0-1.el7_1
Error: Package: python2-docker-3.3.0-1.el7.noarch (centos-openshift-origin-testing-cbs)
           Requires: python-pyOpenSSL
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest

I'm connected to **https://buildlogs.centos.org/centos/7/paas/x86_64/openshift-origin310/**

Full output here:
[https://gist.github.com/isaldarriaga/4203a08fbd5455bc0587c52dfe75205d](https://gist.github.com/isaldarriaga/4203a08fbd5455bc0587c52dfe75205d)

**Comment by moula** [comment]
It works for me

Thank's

2018-08-09 14:39 GMT+02:00 Ivan Saldarriaga <notifications@github.com>:

**Comment by isaldarriaga** [comment]
ok. i'm going to try with a fresh install to confirm.

**Comment by DanyC97** [comment]
fyi i just tagged `pyOpenSSL-17.3.0-3.el7` & `python-requests-2.19.1-3.el7` and it should appear in next 10-15 min in the repo.

Saying that will be good if you can try 2 more fresh installed:

* with the packages i just added => let me know the result
* then i'll remove the packages => see if you get same like @moula in which case there won't be any need for the last 2 rpms

**Comment by isaldarriaga** [comment]
actually ran the installer from fresh install before the new packages to appear in the repo. 

failed in the prerequisites.

"item": "python-docker", 
    "msg": "Error: Package: python2-docker-3.3.0-1.el7.noarch (centos-openshift-origin-testing-cbs)\n           Requires: python-requests >= 2.14.2\n           Available: python-requests-2.6.0-1.el7_1.noarch (base)\n               python-requests = 2.6.0-1.el7_1\nError: Package: python2-docker-3.3.0-1.el7.noarch (centos-openshift-origin-testing-cbs)\n           Requires: python-pyOpenSSL\n",

looks like they are actually needed.

i'll wait an retry. many thanks.

**Comment by isaldarriaga** [comment]
@DanyC97  so far python2-pyOpenSSL-17.3.0-3.el7.noarch.rpm and python2-requests-2.19.1-3.el7.noarch.rpm appear in CBS, but not in buildlogs.

if i change the repo's baseurl from https://buildlogs.centos.org/centos/7/paas/x86_64/openshift-origin310/ to https://cbs.centos.org/repos/paas7-openshift-origin310-testing/x86_64/os/Packages/ and try to `yum -y update` i get:

https://cbs.centos.org/repos/paas7-openshift-origin310-testing/x86_64/os/Packages/repodata/repomd.xml: [Errno 14] HTTPS Error 404 - Not Found

The same error appear when running ansible's prerequisites.yaml too.

will the rpm appear in buildlogs? or better to add repodata folder in CBS?

**Comment by isaldarriaga** [comment]
**update**: 

python2-idna
python2-urllib3

required/missing

using: https://buildlogs.centos.org/centos/7/paas/x86_64/openshift-origin310/

full output:
[https://gist.github.com/isaldarriaga/1c3bcc0cb6a470b42d1a48e021b741e3](https://gist.github.com/isaldarriaga/1c3bcc0cb6a470b42d1a48e021b741e3)

**Comment by DanyC97** [comment]
> so far python2-pyOpenSSL-17.3.0-3.el7.noarch.rpm and python2-requests-2.19.1-3.el7.noarch.rpm appear in CBS, but not in buildlogs.

they are now

> python2-idna
python2-urllib3

will add them soon

**Comment by DanyC97** [comment]
@isaldarriaga here we go: `python-idna-2.5-1.el7` & `python-urllib3-1.16-2.el7` added to the repo, it should appear in next 10-30 min.

any chance you can give it another go and let me know how it goes ? thx

**Comment by isaldarriaga** [comment]
yes, this time I'll add the [lb] (haproxy) to see if more packages are needed or not.

**Comment by isaldarriaga** [comment]
"item": "python-docker", 
    "msg": "Error: Package: python2-urllib3-1.16-2.el7.noarch (centos-openshift-origin-testing-cbs)\n           Requires: python-pysocks\n", 
    "rc": 1,

**Comment by DanyC97** [comment]
@isaldarriaga sorted, `python-pysocks-1.5.6-3.el7` added, give it 10-15 min pls to show up

**Comment by isaldarriaga** [comment]
prerequisites passed. running playbooks/deploy_cluster.yml now.

**Comment by DanyC97** [comment]
@isaldarriaga thank you v much for not giving up and work with me on finding all missing dependencies

**Comment by isaldarriaga** [comment]
@DanyC97 



{
    "attempts": 3, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "allow_downgrade": false, 
            "bugfix": false, 
            "conf_file": null, 
            "disable_gpg_check": false, 
            "disable_plugin": [], 
            "disablerepo": null, 
            "enable_plugin": [], 
            "enablerepo": null, 
            "exclude": null, 
            "install_repoquery": true, 
            "installroot": "/", 
            "list": null, 
            "name": [
                "atomic"
            ], 
            "security": false, 
            "skip_broken": false, 
            "state": "present", 
            "update_cache": false, 
            "update_only": false, 
            "validate_certs": true
        }
    }, 
    "item": "atomic", 
    "msg": "Error: Package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64 (extras)\n           Requires: python-docker-py >= 1.7.2-1\n           Available: docker-python-1.4.0-115.el7.x86_64 (extras)\n               python-docker-py = 1.4.0-115.el7\n           Available: python-docker-py-1.10.6-3.el7.noarch (extras)\n               python-docker-py = 1.10.6-3.el7\n           Available: python-docker-py-1.10.6-4.el7.noarch (extras)\n               python-docker-py = 1.10.6-4.el7\n", 
    "rc": 1, 
    "results": [
        "Loaded plugins: fastestmirror\nLoading mirror speeds from cached hostfile\n * base: centos.mirror.lstn.net\n * extras: mirror.hostduplex.com\n * updates: centos.mirror.lstn.net\nResolving Dependencies\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n--> Processing Dependency: runc for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Processing Dependency: python-dateutil for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Processing Dependency: gomtree for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n---> Package gomtree.x86_64 0:0.5.0-0.2.git16da0f8.el7 will be installed\n---> Package python-dateutil.noarch 0:1.5-7.el7 will be installed\n---> Package runc.x86_64 0:1.0.0-27.rc5.dev.git4bb1fe4.el7 will be installed\n--> Processing Dependency: criu for package: runc-1.0.0-27.rc5.dev.git4bb1fe4.el7.x86_64\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n---> Package criu.x86_64 0:3.5-4.el7 will be installed\n--> Processing Dependency: libprotobuf-c.so.1(LIBPROTOBUF_C_1.0.0)(64bit) for package: criu-3.5-4.el7.x86_64\n--> Processing Dependency: libprotobuf-c.so.1()(64bit) for package: criu-3.5-4.el7.x86_64\n--> Processing Dependency: libnet.so.1()(64bit) for package: criu-3.5-4.el7.x86_64\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n---> Package libnet.x86_64 0:1.1.6-7.el7 will be installed\n---> Package protobuf-c.x86_64 0:1.0.2-3.el7 will be installed\n--> Finished Dependency Resolution\n You could try using --skip-broken to work around the problem\n You could try running: rpm -Va --nofiles --nodigest\n"
    ]
}

**Comment by DanyC97** [comment]
@isaldarriaga now this one is confusing me ...
On which operating system are you deploying OpenShift ? i assumed is centOS in which case atomic pkg should not be installed.

Can you pls confirm ?

**Comment by DanyC97** [comment]
@isaldarriaga so few things:

1) you don't need to switch to tags, there is an rpm for ansible `openshift-ansible-3.10.27-1.git.0.d5723a3.el7.noarch.rpm` [here](http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin310/)
2) thank you for pointing to the other issue, the prob i see is that http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin/ has not been updated except http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin310/ -> i'll fix that.

please note that all the other missing python-* packages were pushed to mirror.centos.org and that will appear Monday morning after 09:00 UTC (a scheduled job needs to run to sign and publish the packages)

**Comment by DanyC97** [comment]
@isaldarriaga can you please confirm if your yum repo is setup like i mentioned [here](https://github.com/openshift/openshift-ansible/issues/7794#issuecomment-412306627)  ? 

i just want to see if you get same output like @danscarf

**Comment by DanyC97** [comment]
> i noticed Ansible creates this file during installation: 

oh !! thank you for this vital info, i'll fire a PR and fix this too in addition to what i mentioned in my [previous](https://github.com/openshift/openshift-ansible/issues/8399#issuecomment-412304263) comment

**Comment by vital-st1x** [comment]
Hi,

I am afraid that this patch is not enough to resolve this problem.

1 ) https://github.com/openshift/openshift-ansible/blob/release-3.10/roles/openshift_repos/tasks/centos_repos.yml#L22 - The file with name CentOS-OpenShift-Origin31.repo.j2 (not Origin310) should be taken.

2) But because {{ openshift_version }} is not defined in this line the file CentOS-OpenShift-Origin.repo.j2 will be used. I suppose the role version should be before role repo
https://github.com/openshift/openshift-ansible/blob/release-3.10/playbooks/init/main.yml#L27-L35

Please correct me if I'm wrong
Thanks

**Comment by DanyC97** [comment]
@vital-st1x please comment directly on the PR and not on this issue.

Trying to keep the history of all the comments close to the initial problem of missing rpms.
I'll look at your comments and respond on the PR itself, thanks !

**Comment by isaldarriaga** [comment]
@DanyC97 i patched my local openshift-ansible repo with the changes that you performed in the PR. 
I also removed any origin's .repo file from the servers (fresh machines setup).

I can tell the new code worked! i.e. the right /etc/yum.repos.d/CentOS-OpenShift-Origin310.repo file was created and pre-requisites ran correctly.

looking forward they become available in the branch after approval/tests/commit.

now I'm running deploy_cluster.yaml. 

so far so good. I'll keep you posted of the final results on this part.

**Comment by isaldarriaga** [comment]
@DanyC97 i got this:

TASK [openshift_node : install needed rpm(s)] *****************************************************************************************************************
task path: /home/ivan/convergence.cloud/dev-ops/kube-prod/origin/openshift-ansible-v3.10/roles/openshift_node/tasks/install_rpms.yml:2

...

{
    "attempts": 3, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "allow_downgrade": false, 
            "bugfix": false, 
            "conf_file": null, 
            "disable_gpg_check": false, 
            "disable_plugin": [], 
            "disablerepo": null, 
            "enable_plugin": [], 
            "enablerepo": null, 
            "exclude": null, 
            "install_repoquery": true, 
            "installroot": "/", 
            "list": null, 
            "name": [
                "atomic"
            ], 
            "security": false, 
            "skip_broken": false, 
            "state": "present", 
            "update_cache": false, 
            "update_only": false, 
            "validate_certs": true
        }
    }, 
    "item": "atomic", 
    "msg": "Error: Package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64 (extras)\n           Requires: python-docker-py >= 1.7.2-1\n           Available: docker-python-1.4.0-115.el7.x86_64 (extras)\n               python-docker-py = 1.4.0-115.el7\n           Available: python-docker-py-1.10.6-3.el7.noarch (extras)\n               python-docker-py = 1.10.6-3.el7\n           Available: python-docker-py-1.10.6-4.el7.noarch (extras)\n               python-docker-py = 1.10.6-4.el7\n", 
    "rc": 1, 
    "results": [
        "Loaded plugins: fastestmirror\nLoading mirror speeds from cached hostfile\n * base: mirrors.xmission.com\n * extras: repos-lax.psychz.net\n * updates: repos.lax.quadranet.com\nResolving Dependencies\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n--> Processing Dependency: runc for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Processing Dependency: python-dateutil for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Processing Dependency: gomtree for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n---> Package gomtree.x86_64 0:0.5.0-0.2.git16da0f8.el7 will be installed\n---> Package python-dateutil.noarch 0:1.5-7.el7 will be installed\n---> Package runc.x86_64 0:1.0.0-27.rc5.dev.git4bb1fe4.el7 will be installed\n--> Processing Dependency: criu for package: runc-1.0.0-27.rc5.dev.git4bb1fe4.el7.x86_64\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n---> Package criu.x86_64 0:3.5-4.el7 will be installed\n--> Processing Dependency: libprotobuf-c.so.1(LIBPROTOBUF_C_1.0.0)(64bit) for package: criu-3.5-4.el7.x86_64\n--> Processing Dependency: libprotobuf-c.so.1()(64bit) for package: criu-3.5-4.el7.x86_64\n--> Processing Dependency: libnet.so.1()(64bit) for package: criu-3.5-4.el7.x86_64\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n---> Package libnet.x86_64 0:1.1.6-7.el7 will be installed\n---> Package protobuf-c.x86_64 0:1.0.2-3.el7 will be installed\n--> Finished Dependency Resolution\n You could try using --skip-broken to work around the problem\n You could try running: rpm -Va --nofiles --nodigest\n"
    ]
}

Seems like the atomic dependency is not handled correctly at this point because i explicitly refused its use in the [inventory](https://gist.github.com/isaldarriaga/9d5405d7ab3db26aeb26faa3cdc21c3e) file:
`containerized=false` as node and [OSEv3:vars] attributes

and as far as i know it's deprecated in 3.10 too.

i ran it in fresh install twice, getting the same error. 
Also made sure the git branch was correct (release-3.10, no tags)

by the other hand the inventory includes:

`openshift_release=v3.10`
`openshift_repos_enable_testing=true`

which generates:

`$ cat /etc/yum.repos.d/CentOS-OpenShift-Origin310.repo`

`[centos-openshift-origin310]`
`name=CentOS OpenShift Origin`
`baseurl=http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin310/`
`enabled=1`
`gpgcheck=1`
`gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-SIG-PaaS`

`[centos-openshift-origin310-testing]`
`name=CentOS OpenShift Origin Testing`
`baseurl=http://buildlogs.centos.org/centos/7/paas/x86_64/openshift-origin310/`
`enabled=1`
`gpgcheck=0`
`gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-SIG-PaaS`

`[centos-openshift-origin310-debuginfo]`
`name=CentOS OpenShift Origin DebugInfo`
`baseurl=http://debuginfo.centos.org/centos/7/paas/x86_64/`
`enabled=0`
`gpgcheck=1`
`gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-SIG-PaaS`

`[centos-openshift-origin310-source]`
`name=CentOS OpenShift Origin Source`
`baseurl=http://vault.centos.org/centos/7/paas/Source/openshift-origin310/`
`enabled=0`
`gpgcheck=1`
`gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-SIG-PaaS[ivan@asus-rog openshift-ansible-v3.10]`

i attached the output at verbosity 8:

[ansible.log](https://github.com/openshift/openshift-ansible/files/2285265/ansible.zip)

I'll take a look to this output tomorrow. if you find something let me know.

thanks.

**Comment by DanyC97** [comment]
@isaldarriaga this was reported by someone else too on the dev mailer. I would suggest open a new issue as this is very likely a bug (atomic should not be installed)

**Comment by aramalipoor** [comment]
After checking out `openshift-ansible-3.10.27-1` tag this `atomic` issue didn't happen.

**Comment by DanyC97** [comment]
@Reamer my pleasure ;) . I still need to publish some rpms into common sub dir but we hit a minor issue which i'm working on with CentOS infra folks.

regarding using or not the rpm, is up to you however i'd suggest if you using git to at least stick to the tag and not tip of the branch. Things are fluid as you can see so you might not always get the same result

**Comment by DanyC97** [comment]
@Reamer @sebile @isaldarriaga fyi the last piece of the puzzle has been sorted, the job run this morning updated the [common repo](http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin/common/?C=M;O=D)

Also tomorrow morning after 09:00 UTC you should see a new rpm `openshift-ansible-3.10.28-1.git.0.9242c73` which includes the @vrutkovs 's fix around openvswitch rpm which is no longer required.

**Comment by DanyC97** [comment]
After tomorrow's job we should all be aligned and have no rpm issues anymore, hence we could close this issue.

Thank you for your patience while sailing through.

**Comment by Reamer** [comment]
@DanyC97 What's the difference between [openshift-origin/common](http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin/common/) and [openshift-origin](http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin/)?

**Comment by sdodson** [comment]
FYI, I've asked @DanyC97 to untag those packages from the CentOS PaaS SIG repos because they introduce package dependency issues. I'll work on reading up on this issue and proposing an alternate solution and/or set of correct packages to tag into those repos.

**Comment by imranrazakhan** [comment]
i just tried again

OS Version: CentOS 7.5

```
git clone https://github.com/openshift/openshift-ansible.git
cd openshift-ansible && git fetch && git checkout release-3.10 && cd ..
ansible-playbook -i inventory.ini openshift-ansible/playbooks/prerequisites.yml
ansible-playbook -i inventory.ini openshift-ansible/playbooks/deploy_cluster.yml
```

Following is error

    {  
       "attempts":3,
       "changed":false,
       "item":"atomic",
       "msg":"Error: Package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64 (extras)\n           Requi 
    res: python-docker-py >= 1.7.2-1\n           Available: docker-python-1.4.0-115.el7.x86_64 (extras)\n               python-docker-py = 1.4.0-115.el7\n",
       "rc":1,
       "results":[  
          "Loade 
    d plugins: fastestmirror\nLoading mirror speeds from cached hostfile\n * base: ftp.fau.de\n * epel: mirrors.dotsrc.org\n * extras: mirror2.hs-esslingen.de\n * updates: mirror2.hs-e 
    sslingen.de\nResolving Dependencies\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: skopeo >= 1: 
    0.1.29-3 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Pr 
    ocessing Dependency: runc for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Processing Dependency: python-dateutil for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n- 
    -> Processing Dependency: ostree for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Processing Dependency: gomtree for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n-- 
    > Running transaction check\n---> Package gomtree.x86_64 0:0.5.0-0.2.git16da0f8.el7 will be installed\n---> Package ostree.x86_64 0:2018.5-1.el7 will be installed\n--> Processing D 
    ependency: libarchive.so.13()(64bit) for package: ostree-2018.5-1.el7.x86_64\n---> Package python-dateutil.noarch 0:1.5-7.el7 will be installed\n---> Package python-docker-py.noarc 
    h 0:1.10.6-4.el7 will be installed\n---> Package runc.x86_64 0:1.0.0-27.rc5.dev.git4bb1fe4.el7 will be installed\n--> Processing Dependency: criu for package: runc-1.0.0-27.rc5.dev 
    .git4bb1fe4.el7.x86_64\n---> Package skopeo.x86_64 1:0.1.31-1.dev.gitae64ff7.el7.centos will be installed\n--> Running transaction check\n---> Package criu.x86_64 0:3.5-4.el7 will 
    be installed\n--> Processing Dependency: libprotobuf-c.so.1(LIBPROTOBUF_C_1.0.0)(64bit) for package: criu-3.5-4.el7.x86_64\n--> Processing Dependency: libprotobuf-c.so.1()(64bit) f 
    or package: criu-3.5-4.el7.x86_64\n--> Processing Dependency: libnet.so.1()(64bit) for package: criu-3.5-4.el7.x86_64\n---> Package libarchive.x86_64 0:3.1.2-10.el7_2 will be insta 
    lled\n--> Running transaction check\n---> Package libnet.x86_64 0:1.1.6-7.el7 will be installed\n---> Package protobuf-c.x86_64 0:1.0.2-3.el7 will be installed\nRemoving python-doc 
    ker-py.noarch 0:1.10.6-4.el7 - u due to obsoletes from installed python2-docker-3.3.0-1.el7.noarch\n--> Restarting Dependency Resolution with new changes.\n--> Running transaction 
    check\n---> Package python-docker-py.noarch 0:1.10.6-4.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7. 
    x86_64\n--> Running transaction check\n---> Package python-docker-py.noarch 0:1.10.6-3.el7 will be installed\nRemoving python-docker-py.noarch 0:1.10.6-3.el7 - u due to obsoletes f 
    rom installed python2-docker-3.3.0-1.el7.noarch\n--> Restarting Dependency Resolution with new changes.\n--> Running transaction check\n---> Package python-docker-py.noarch 0:1.10. 
    6-3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Finished Dependency Resolution\n You co 
    uld try using --skip-broken to work around the problem\n You could try running: rpm -Va --nofiles --nodigest\n"
       ]
    }

**Comment by joshainglis** [comment]
@imranrazakhan a quick fix for that issue is to remove `- atomic` from the `default_r_openshift_node_image_prep_packages` variable in `roles/openshift_node/defaults/main.yml`

**Comment by DanyC97** [comment]
@imranrazakhan sorry for the issue you got, we are basically back to square one after we rollback the changes i made.

If you don't mind asking, any chance you can post here or in a gist the cat out of all the files you have in your `/etc/yum.repos.d/` dir?

**Comment by DanyC97** [comment]
@imranrazakhan in your output i can see

```
Package python-docker-py.noarch 0:1.10.6-4.el7 will be installed
--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64
--> Running transaction check
---> Package python-docker-py.noarch 0:1.10.6-3.el7 will be installed\nRemoving python-docker-py.noarch 0:1.10.6-3.el7 - u due to obsoletes from installed python2-docker-3.3.0-1.el7.noarch
--> Restarting Dependency Resolution with new changes.
--> Running transaction check
---> Package python-docker-py.noarch 0:1.10.6-3.el7 will be installed
--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64
--> Finished Dependency Resolution
You could try using --skip-broken to work around the problem
You could try running: rpm -Va --nofiles --nodigest"

```

Now i just done myself a quick test to see the dependecies for atomic pkg and so see http://pastebin.centos.org/1504251/

As such suggest you start from a clean env, or uninstall the the previous pkg and try again.

@isaldarriaga @Reamer you guys able to help and try a fresh deployment on a clean env and let me know the result  ?

**Comment by alv91** [comment]
@imranrazakhan check if you have python2-docker installed, then remove it. 

>     70 | remove python2-docker    | 2018-08-14 16:31 | Erase          |    1  

Atomic has dependence python-docker-py >= 1.7.2-1, in my env it was blocked by this package.

**Comment by DanyC97** [comment]
@alv91 indeed atomic depends on `python-docker-py` however is not **pinned** to a specific version. As shown in my previous comment, it pulls down the latest `python-docker-py` version if there is nothing present.

**Comment by alv91** [comment]
@DanyC97 Yes, but i had to remove this package to allow reinstallation atomic package. I cant find also any installation package like `python2-docker` in openshift-ansible repository. 

```
# rpm -i python2-docker-3.3.0-1.el7.noarch.rpm 
error: Failed dependencies:
	python-docker-py < 2 is obsoleted by python2-docker-3.3.0-1.el7.noarch
```

**Comment by DanyC97** [comment]
@alv91 correct because we had to remove the package yesterday evening. And so the expectation is that everyone who already have a non pristine deployment or an env where the `python2-docker-3.3.0-1.el7.noarch.rpm` & `python-docker-py >= 1.7.2-1` are present to _uninstall it_

**Comment by isaldarriaga** [comment]
@DanyC97 

fresh install, this time:

- manually copied [/etc/yum.repos.d/CentOS-OpenShift-Origin310.repo](https://gist.github.com/isaldarriaga/a2c98d74b88967b58fd0f77c6c2b7d39) to all hosts (looks like the  [MERGE](https://github.com/openshift/openshift-ansible/commit/e272e0caab8b0b5bd670d67e7de52859bb7012ea) code of [PR9548](https://github.com/openshift/openshift-ansible/pull/9548#pullrequestreview-145576505) is not doing this correctly. could you please verify this?)
- not using sdn but flannel (errors during deploy prevent etcd, infra, and worker nodes to start/join openshift cluster with [empty /etc/cni/net.d/](https://github.com/openshift/openshift-ansible/issues/7967))
- not using firewalld but iptables (firewall rule application errors detected possibly causing sdn not to start)
- removed firewalld service from hosts (service in failed stated)
- disabled firewall and rules at the virtualization level (just in case)
- made sure the hosts have the following [packages](https://docs.okd.io/3.10/install/host_preparation.html#installing-base-packages) pre-installed: `wget git net-tools bind-utils yum-utils iptables-services bridge-utils bash-completion kexec-tools sos psacct`
- not using dedicated etcd hosts ('cause noticed masters are forced to include etcd no matter they are not part of [etcd], so had x6 etcd during install. only 3 needed)
- synchronized server [clocks](https://gist.github.com/isaldarriaga/4c968ef390519df2f4323e55100037a9)
- made sure no system services are failing (BTW, kdump / kexec-tools requires [2GB](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/kernel_administration_guide/kernel_crash_dump_guide#sect-kdump-memory-requirements) RAM as minimum for this service to be up)
- increased all `retries: n` and `delay: m` to 100 so i have room to restart NetworkManager and dnsmasq services during installation (master to master connectivity is lost during httpd-tools rpm setup)
- testing both `3.10` branch and `3.10.28-1` tag

test results:

```
$ BRANCH="3.10" && \
mkdir --parents "openshift-ansible-v$BRANCH" && \
git clone https://github.com/openshift/openshift-ansible "openshift-ansible-v$BRANCH" && \
cd "openshift-ansible-v$BRANCH" && \
git checkout release-$BRANCH && \
echo && echo "Current version is:" && \
git describe && \
cd ..
```

```
Cloning into 'openshift-ansible-v3.10'...
remote: Counting objects: 119843, done.
remote: Total 119843 (delta 0), reused 0 (delta 0), pack-reused 119842
Receiving objects: 100% (119843/119843), 31.00 MiB | 1.84 MiB/s, done.
Resolving deltas: 100% (75124/75124), done.
Branch 'release-3.10' set up to track remote branch 'release-3.10' from 'origin'.
Switched to a new branch 'release-3.10'

Current version is:
openshift-ansible-3.10.29-1-2-g8e9cd4a1b
```

`ansible-playbook -vvv -i /path/to/inventory.ini /path/to/playbooks/prerequisites.yml`

```
TASK [openshift_node : install needed rpm(s)] 
task path: /home/ivan/convergence.cloud/dev-ops/kube-prod/origin/openshift-ansible-v3.10/roles/openshift_node/tasks/install_rpms.yml:2
...
FAILED - RETRYING: install needed rpm(s) (1 retries left).Result was:
...
"item": "atomic", 
    "msg": "Error: Package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64 (extras)\n           Requires: python-docker-py >= 1.7.2-1\n           Available: docker-python-1.4.0-115.el7.x86_64 (extras)\n               python-docker-py = 1.4.0-115.el7\n           Available: python-docker-py-1.10.6-3.el7.noarch (extras)\n               python-docker-py = 1.10.6-3.el7\n           Available: python-docker-py-1.10.6-4.el7.noarch (extras)\n               python-docker-py = 1.10.6-4.el7\n"
```

I'll checkout `3.10.28-1` tag and post results later.

**Comment by isaldarriaga** [comment]
getting tag code:

```BRANCH="3.10" && TAG="3.10.28-1" && \
mkdir --parents "openshift-ansible-v$TAG" && \
git clone https://github.com/openshift/openshift-ansible "openshift-ansible-v$TAG" && \
cd "openshift-ansible-v$TAG" && \
git fetch --all --tags --prune && \
git checkout tags/openshift-ansible-$TAG -b release-$BRANCH
echo && echo "Current version is:" && \
git describe --tags && \
cd ..
```

`ansible-playbook -vvv -i /path/to/inventory.ini /path/to/playbooks/prerequisites.yml`

exact same output on a fresh install:

```
{
    "attempts": 3, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "allow_downgrade": false, 
            "bugfix": false, 
            "conf_file": null, 
            "disable_gpg_check": false, 
            "disable_plugin": [], 
            "disablerepo": null, 
            "enable_plugin": [], 
            "enablerepo": null, 
            "exclude": null, 
            "install_repoquery": true, 
            "installroot": "/", 
            "list": null, 
            "name": [
                "atomic"
            ], 
            "security": false, 
            "skip_broken": false, 
            "state": "present", 
            "update_cache": false, 
            "update_only": false, 
            "validate_certs": true
        }
    }, 
    "item": "atomic", 
    "msg": "Error: Package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64 (extras)\n           Requires: python-docker-py >= 1.7.2-1\n           Available: docker-python-1.4.0-115.el7.x86_64 (extras)\n               python-docker-py = 1.4.0-115.el7\n           Available: python-docker-py-1.10.6-3.el7.noarch (extras)\n               python-docker-py = 1.10.6-3.el7\n           Available: python-docker-py-1.10.6-4.el7.noarch (extras)\n               python-docker-py = 1.10.6-4.el7\n", 
    "rc": 1, 
    "results": [
        "Loaded plugins: fastestmirror\nLoading mirror speeds from cached hostfile\n * base: repos.lax.quadranet.com\n * extras: mirrors.xmission.com\n * updates: centos.sonn.com\nResolving Dependencies\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n--> Processing Dependency: runc for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Processing Dependency: python-dateutil for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Processing Dependency: gomtree for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n---> Package gomtree.x86_64 0:0.5.0-0.2.git16da0f8.el7 will be installed\n---> Package python-dateutil.noarch 0:1.5-7.el7 will be installed\n---> Package runc.x86_64 0:1.0.0-27.rc5.dev.git4bb1fe4.el7 will be installed\n--> Processing Dependency: criu for package: runc-1.0.0-27.rc5.dev.git4bb1fe4.el7.x86_64\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n---> Package criu.x86_64 0:3.5-4.el7 will be installed\n--> Processing Dependency: libprotobuf-c.so.1(LIBPROTOBUF_C_1.0.0)(64bit) for package: criu-3.5-4.el7.x86_64\n--> Processing Dependency: libprotobuf-c.so.1()(64bit) for package: criu-3.5-4.el7.x86_64\n--> Processing Dependency: libnet.so.1()(64bit) for package: criu-3.5-4.el7.x86_64\n--> Running transaction check\n---> Package atomic.x86_64 1:1.22.1-22.git5a342e3.el7 will be installed\n--> Processing Dependency: python-docker-py >= 1.7.2-1 for package: 1:atomic-1.22.1-22.git5a342e3.el7.x86_64\nPackage python-docker-py-1.10.6-4.el7.noarch is obsoleted by python2-docker-3.3.0-1.el7.noarch which is already installed\n---> Package libnet.x86_64 0:1.1.6-7.el7 will be installed\n---> Package protobuf-c.x86_64 0:1.0.2-3.el7 will be installed\n--> Finished Dependency Resolution\n You could try using --skip-broken to work around the problem\n You could try running: rpm -Va --nofiles --nodigest\n"
    ]
}

```

**Comment by sdodson** [comment]
@DanyC97 Can we tag this into our repos? This is the same version that's shipped in OCP and it fulfills all of the package requirements of atomic.
http://cbs.centos.org/koji/buildinfo?buildID=17580

**Comment by sdodson** [comment]
@isaldarriaga If you install the package from http://cbs.centos.org/koji/buildinfo?buildID=17580 are all the dependencies of atomic fulfilled?

**Comment by isaldarriaga** [comment]
@sdodson I haven't tried that option yet but could in a next run.

i already did what @joshainglis [suggested](https://github.com/openshift/openshift-ansible/issues/8399#issuecomment-413474814), to comment/remove "atomic" from default_r_openshift_node_image_prep_packages.

so far so good, this approach let the prerequisites continue to further steps.

I'll keep you posted. waiting for results yet.

**Comment by DanyC97** [comment]
@sdodson i've tagged the the [rpm](http://cbs.centos.org/koji/buildinfo?buildID=17580) into [candidate repo](https://cbs.centos.org/repos/paas7-openshift-origin310-candidate/x86_64/os/Packages/?C=M;O=D) as well as [testing repo](https://buildlogs.centos.org/centos/7/paas/x86_64/openshift-origin310/?C=M;O=D)

Before tagging into -release i'll wait for @isaldarriaga to confirm if it worked or not as i don't want to go again through the process of removing rpms.

@isaldarriaga regarding 

> manually copied /etc/yum.repos.d/CentOS-OpenShift-Origin310.repo to all hosts (looks like the MERGE code of PR9548 is not doing this correctly. could you please verify this?)

can you share more info why you think is not doing the right thing ?

**Comment by isaldarriaga** [comment]
@sdodson 
just getting a couple of non related errors to this issue.

1. Hosts:    prod-lb-01.example.com
     Play:     Initialize cluster facts
     Task:     Gather Cluster facts
     Message:  MODULE FAILURE

  2. Hosts:    prod-master-01.example.com
     Play:     Create Hosted Resources - router
     Task:     Get the certificate contents for router
     Message:  |lib_utils_oo_select_keys failed expects to filter on a dict or object

I'll work in a solution, retry and let you know.

@DanyC97 

> can you share more info why you think is not doing the right thing ?

sure, no matter i've set `openshift_release=v3.10` and `openshift_repos_enable_testing=true` in the inventory, the `/etc/yum.repos.d/CentOS-OpenShift-Origin310.repo` file does not appear in the server during prerequisites nor cluster_deploy steps, thence there's no valid source to pull the rpms from. This happen in the current v3.10 branch (`$ git describe => openshift-ansible-3.10.29-1-2-g8e9cd4a1b`) which already has the PR changes merged. 

I was able to [successfully test this part before](https://github.com/openshift/openshift-ansible/issues/8399#issuecomment-412594943), i think in 3.10-27-1, by locally patching roles/openshift_repos/tasks/[centos_repos.yml](https://gist.github.com/isaldarriaga/23d586c6253d5e8347c18d74447318cb) and adding the roles/openshift_repos/templates/[CentOS-OpenShift-Origin310.repo.j2](https://gist.github.com/isaldarriaga/ba2a58795aa2b8400217b013bbc24359) file. With the patch, ansible did copy the [CentOS-OpenShift-Origin310.repo](https://gist.github.com/isaldarriaga/46136d72af4d8c55564de368ea964073) to the servers. see links with its respective content.

Maybe here there's a clue: I noticed centos_repos.yml is no longer part of the PR nor the merge, which initially included this change: `CentOS-OpenShift-Origin{{ ((openshift_release | default('')).split('.') | join(''))[0:3] }}.repo.j2`, but even if i patch the current branch with that sentence i cannot get the CentOS-OpenShift-Origin310.repo generated in the servers either. so that's the reason i decided to manually copy the CentOS-OpenShift-Origin310.repo file to the servers before the test.

hope this help to clarify.

**Comment by DanyC97** [comment]
@isaldarriaga if you don't mind, any chance you can try a fresh deployment using the testing repo while leave the atomic rpm in and let me know if you happy or not ? 

reason for that is because i'd like then to push the change to -release and try to close this issue which is getting closer to 100 comments.

I hope we can put to rest the rpms issue and then focus on the other issue you just mentioned.

**Comment by vrutkovs** [comment]
>I don't understand why installing any RPM packages on the masters/nodes when containerized=true

Containerized install has been deprecated in 3.10. It won't do anything now.

Initial reporter has `openshift_image_tag=v3.9.0` which would certainly break

This issue is becoming a collection of 'I also have problem with repo' but it doesn't mean its all caused by the same problem in the code, so I'm going to close it

Please open new issue for particular problem - and don't forget to attach the inventory, exact branch and the output of playbook run with '-vvv'.

**Comment by isaldarriaga** [comment]
@DanyC97 prerequisites were not running in my environment due an error in one of my scripts.  deploy_cluster.yaml was being running instead. i fixed that.

i ran fresh again

`$ git describe`
`openshift-ansible-3.10.29-1-22-gdec10a6d6`

i didn't remove `atomic` from `default_r_openshift_node_image_prep_packages` this time, nor patched `roles/openshift_repos/tasks/centos_repos.yml`

during prerequisites execution:

```
$ ls -al /etc/yum.repos.d/
total 48
drwxr-xr-x.  2 root root  223 Aug 17 04:23 .
drwxr-xr-x. 78 root root 8192 Aug 17 04:21 ..
-rw-r--r--.  1 root root 1664 Aug  2 03:56 CentOS-Base.repo
-rw-r--r--.  1 root root 1309 Aug  2 03:56 CentOS-CR.repo
-rw-r--r--.  1 root root  649 Aug  2 03:56 CentOS-Debuginfo.repo
-rw-r--r--.  1 root root  314 Aug  2 03:56 CentOS-fasttrack.repo
-rw-r--r--.  1 root root  630 Aug  2 03:56 CentOS-Media.repo
-rw-r--r--.  1 root root  887 Aug 17 04:23 CentOS-OpenShift-Origin.repo
-rw-r--r--.  1 root root 1331 Aug  2 03:56 CentOS-Sources.repo
-rw-r--r--.  1 root root 4768 Aug  2 03:56 CentOS-Vault.repo
```

`$ cat /etc/yum.repos.d/CentOS-OpenShift-Origin.repo # master-01`

[output](https://gist.github.com/isaldarriaga/dd1aa94a90f742148b8678b2a73815e0)

**prerequisites ran to completion without errors with this repo file. So the PR in mention was actually ok!**

running deploy_cluster.yaml now.

will re-open if another rpm issue is found.

thanks!
